apps/test-app/frontend/src/components/tree-widget/TreeWidget.tsx

```diff
@@ -170,23 +182,73 @@ export function StatelessTreeWidget(props: Omit<Props, "rulesetId">) {
   const dataProvider = useMemo((): TreeDataProvider => {
     return async (node?: TreeNodeItem): Promise<TreeNodeItem[]> => {
       const parent: HierarchyNode | undefined = node ? (node as any).__internal : undefined;
+      let lastNonGroupingNodeId: string | undefined;
+      // Retrieve Id of last non grouping node
+      if (parent) {
+        let nodeKey = parent.key;
+        if (HierarchyNode.isGroupingNode(parent)) {
+          for (let i = parent.parentKeys.length - 1; i >= 0; --i) {
+            if (!HierarchyNodeKey.isGrouping(parent.parentKeys[i])) {
+              nodeKey = parent.parentKeys[i];
+              break;
+            }
+          }
+        }
+        if (!HierarchyNodeKey.isGrouping(nodeKey)) {
+          if (typeof nodeKey === "string") {
+            lastNonGroupingNodeId = nodeKey;
+          } else {
+            lastNonGroupingNodeId = nodeKey.instanceKeys[0].id;
+          }
+        }
+      }
+      // Get limit that is set for lastNonGroupingNodeId
+      const mapKeyValue = hierarchyLevelSizeLimit.get(lastNonGroupingNodeId);
+      let limit: undefined | number | "unbounded";
+      if (mapKeyValue === undefined) {
+        hierarchyLevelSizeLimit.set(lastNonGroupingNodeId, { limit });
```

Comment:
Do we need to set it here? I don't think it's a good idea to change state from `useMemo` which has a dependency on that state variable.

Reply:
It's a mistake, should not be here at all.
Removed.